proxy_ring_buffer.go

```diff
@@ -0,0 +1,266 @@
+	// sacrificing the time-to-live in favor of being able to write.
+	//
+	// Example: if there is a time-to-live of 30 seconds, the buffer is full
+	// and the oldest item in the ring has been placed a second ago, then there
+	// are 29 seconds remaining for that item. Using this policy you are going
+	// to overwrite no matter what.
+	OverflowPolicyOverwrite OverflowPolicy = 0
+
+	// OverflowPolicyFail
+	// Using this policy the call will fail immediately and the oldest item will
+	// not be overwritten before it is old enough to retire. So this policy
+	// sacrificing the ability to write in favor of time-to-live.
+	//
+	// The advantage of OverflowPolicyFail is that the caller can decide what to do
+	// since it doesn't trap the thread due to backoff.
+	//
+	// Example: if there is a time-to-live of 30 seconds, the buffer is full
+	// and the oldest item in the ring has been placed a second ago, then there
+	// are 29 seconds remaining for that item. Using this policy you are not
+	// going to overwrite that item for the next 29 seconds.
+	OverflowPolicyFail OverflowPolicy = 1
+
+	// ReadResultSetSequenceUnavailable is used when error happened
+	ReadResultSetSequenceUnavailable int64 = -1
+)
+
+func newRingbuffer(p *proxy) (*Ringbuffer, error) {
+	if partitionID, err := p.stringToPartitionID(p.name); err != nil {
+		return nil, err
+	} else {
+		return &Ringbuffer{proxy: p, partitionID: partitionID}, nil
+	}
+}
+
+// Add an item to the tail of the Ringbuffer. If there is space in the Ringbuffer, the call
+// will return the sequence of the written item. If there is no space, it depends on the overflow policy what happens:
+// OverflowPolicy OverflowPolicyOverwrite  we just overwrite the oldest item in the Ringbuffer, and we violate the ttl
+// OverflowPolicy FAIL we return -1. The reason that FAIL exist is to give the opportunity to obey the ttl.
+//
+// This sequence will always be unique for this Ringbuffer instance, so it can be used as a unique id generator if you are
+// publishing items on this Ringbuffer. However, you need to take care of correctly determining an initial id when any node
+// uses the Ringbuffer for the first time. The most reliable way to do that is to write a dummy item into the Ringbuffer and
+// use the returned sequence as initial  id. On the reading side, this dummy item should be discarded. Please keep in mind that
+// this id is not the sequence of the item you are about to publish but from a previously published item. So it can't be used
+// to find that item.
+//
+// Add returns the sequence number of the added item. You can read the added item using this number.
+func (rb *Ringbuffer) Add(ctx context.Context, item interface{}, overflowPolicy OverflowPolicy) (sequence int64, err error) {
+	elementData, err := rb.validateAndSerialize(item)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	request := codec.EncodeRingbufferAddRequest(rb.name, int32(overflowPolicy), elementData)
```

Comment:
How do we validate enum like values in go client? We can check if `overflowPolicy` is in range [0, 1], but we need to update the valid range as well if a new overflowPolicy is added in the future.

Reply:
Go does not support ENUMs as we know from Java.
I did choose for the closest, what is typically used in Go: a type + some constants.
See https://github.com/nitram509/hazelcast-go-client/blob/62089238940f24f70e8c300cf5f265c2e3a91542/proxy_ring_buffer.go#L66

This approach guarantees type safety and thus the compiler checks for values.
Also, this approach can be used in the client code generators. ... similar to enums, such Go snippets can be created.
Then future extensions would be safe.

@utku-caglayan on your concrete example (line 110), in the method signature, is using the type already ... so we have some safety at compile time.

Reply:
I did not understand the type safety provided here. This is valid code: 

```
sequence, err := rb.Add(ctx, "item 1", 123)
```

Reply:
Oops, you're right. I was fallen under the wrong assumption of type safety here,
but I realized type conversion happens :/

Reply:
Serkan, thanks for pointing out the missing type-safety.
On a more general level...

Would you consider following the idea of using type safety, by e.g. using interfaces?
This way, one could not provide an arbitrary int value.

Or would you rather consider explicit range checks and no further implementations to enforce type checks?

I could provide an example if you want?

Reply:
I think @yuce will know what to do better than me. I don't know what we do in the client code and not so familiar with golang as well.

Reply:
Since this value is directly sent to the server-side, it would be better to check it. I think it's enough to check whether the policy is either `OverflowPolicyOverwrite` or `OverflowPolicyFail`, which are the only valid values. If they are not valid, `ErrIllegalArgument` may be returned, something like:
```
if int(overflowPolicy) != 0 || int(overflowPolicy) != 1 {
	msg := fmt.Sprintf("invalid overflow policy: %d", overflowPolicy)
	return 0, hzcerrors.NewIllegalArgumentError(msg, nil)	
}
```
I had to cast the `overflowPolicy` to `int`, otherwise GoLand is trying to simplify the condition, and replaces it with `true`.

Reply:
Done.
99fa52328e90a44e5ef1fcaade10b769e00fd608

Reply:
@yuce please allow me some feedback:
```if int(overflowPolicy) != 0 || int(overflowPolicy) != 1 {```
will always return true (bool arethmetics).
The correct phrase is
```if overflowPolicy != OverflowPolicyFail && overflowPolicy != OverflowPolicyOverwrite```